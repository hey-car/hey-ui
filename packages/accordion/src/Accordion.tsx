import React from 'react';
import cn from 'classnames';

import Collapse from '@heycar-uikit/collapse';
import { ChevronDown } from '@heycar-uikit/icons';

import { AccordionProps } from './Accordion.types';

import styles from './styles/default.module.css';

const Accordion = React.forwardRef<HTMLElement, AccordionProps>(
  (
    {
      children,
      title,
      open = false,
      Component = 'div',
      disabled = false,
      dataTestId,
      className,
      onExpandedChange,
      onTransitionEnd,
      ...restProps
    },
    ref,
  ) => {
    const [isOpen, setIsOpen] = React.useState(open);
    const classNames = cn(
      styles.accordion,
      {
        [styles.accordionOpen]: isOpen,
      },
      className,
    );
    const headerClassNames = cn(styles.header, {
      [styles.headerOpen]: isOpen,
      [styles.headerDisabled]: disabled,
    });
    //const accordionId = `collapsible-${Date.now()}`;
    const handleHeaderButtonChange = (
      event:
        | React.MouseEvent<HTMLElement>
        | React.KeyboardEvent<HTMLDivElement>,
    ) => {
      if (disabled) return event.preventDefault();

      setIsOpen(!isOpen);
      if (onExpandedChange) {
        onExpandedChange(!isOpen);
      }
    };
    const tabIndex = disabled ? -1 : 0;

    return (
      <Component
        className={classNames}
        data-accordion-component="Accordion"
        data-test-id={dataTestId}
        ref={ref}
        {...restProps}
      >
        <div
          // aria-controls={accordionId}
          aria-expanded={isOpen}
          className={headerClassNames}
          data-accordion-component="AccordionItemButton"
          onClick={handleHeaderButtonChange}
          onKeyPress={handleHeaderButtonChange}
          role="button"
          tabIndex={tabIndex}
        >
          {title}
          <div className={styles.headerArrow}>
            <ChevronDown aria-hidden="true" color="inherit" />
          </div>
        </div>
        <Collapse
          // aria-labelledby={accordionId}
          className={styles.body}
          onTransitionEnd={onTransitionEnd}
          role="region"
          data-accordion-component="AccordionItemPanel"
          // id={accordionId}
          open={isOpen}
        >
          {children}
        </Collapse>
      </Component>
    );
  },
);

export default Accordion;
