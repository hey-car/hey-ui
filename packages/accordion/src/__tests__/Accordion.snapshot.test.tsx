import React from 'react';
import { render } from '@testing-library/react';

import Accordion from '../Accordion';

const accordionTitle = 'title';
const accordionBodyText =
  ' Youâ€™ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is';
const dateNow = 1661419668701;

describe('Accordion Snapshots tests', () => {
  beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(dateNow);
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should match snapshot', () => {
    expect(
      render(<Accordion title={accordionTitle}>{accordionBodyText}</Accordion>),
    ).toMatchSnapshot();
  });

  it('should render accordion title with HTML tag', () => {
    expect(
      render(
        <Accordion title={<h1>{accordionTitle}</h1>}>
          {accordionBodyText}
        </Accordion>,
      ),
    ).toMatchSnapshot();
  });

  it('should render accordion `disabled`', () => {
    expect(
      render(
        <Accordion disabled={true} title={accordionTitle}>
          {accordionBodyText}
        </Accordion>,
      ),
    ).toMatchSnapshot();
  });

  it('should render accordion `open=true` by default', () => {
    expect(
      render(
        <Accordion open={true} title={accordionTitle}>
          {accordionBodyText}
        </Accordion>,
      ),
    ).toMatchSnapshot();
  });
});
