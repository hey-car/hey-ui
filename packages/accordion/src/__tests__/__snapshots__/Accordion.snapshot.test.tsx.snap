// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Snapshots tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="accordion"
        data-accordion-component="Accordion"
      >
        <div
          aria-expanded="false"
          class="header"
          data-accordion-component="AccordionItemButton"
          role="button"
          tabindex="0"
        >
          title
          <div
            class="headerArrow"
          >
            <svg
              aria-hidden="true"
              class="svgIcon inherit"
              data-test-id="ChevronDownIcon"
              focusable="false"
              style="font-size: 24px;"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="collapse body"
          data-accordion-component="AccordionItemPanel"
          role="region"
          style="height: 0px;"
        >
           You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="accordion"
      data-accordion-component="Accordion"
    >
      <div
        aria-expanded="false"
        class="header"
        data-accordion-component="AccordionItemButton"
        role="button"
        tabindex="0"
      >
        title
        <div
          class="headerArrow"
        >
          <svg
            aria-hidden="true"
            class="svgIcon inherit"
            data-test-id="ChevronDownIcon"
            focusable="false"
            style="font-size: 24px;"
            viewBox="0 0 24 24"
          >
            <path
              clip-rule="evenodd"
              d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <div
        class="collapse body"
        data-accordion-component="AccordionItemPanel"
        role="region"
        style="height: 0px;"
      >
         You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Accordion Snapshots tests should render accordion \`disabled\` 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="accordion"
        data-accordion-component="Accordion"
      >
        <div
          aria-expanded="false"
          class="header headerDisabled"
          data-accordion-component="AccordionItemButton"
          role="button"
          tabindex="-1"
        >
          title
          <div
            class="headerArrow"
          >
            <svg
              aria-hidden="true"
              class="svgIcon inherit"
              data-test-id="ChevronDownIcon"
              focusable="false"
              style="font-size: 24px;"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="collapse body"
          data-accordion-component="AccordionItemPanel"
          role="region"
          style="height: 0px;"
        >
           You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="accordion"
      data-accordion-component="Accordion"
    >
      <div
        aria-expanded="false"
        class="header headerDisabled"
        data-accordion-component="AccordionItemButton"
        role="button"
        tabindex="-1"
      >
        title
        <div
          class="headerArrow"
        >
          <svg
            aria-hidden="true"
            class="svgIcon inherit"
            data-test-id="ChevronDownIcon"
            focusable="false"
            style="font-size: 24px;"
            viewBox="0 0 24 24"
          >
            <path
              clip-rule="evenodd"
              d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <div
        class="collapse body"
        data-accordion-component="AccordionItemPanel"
        role="region"
        style="height: 0px;"
      >
         You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Accordion Snapshots tests should render accordion \`open=true\` by default 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="accordion accordionOpen"
        data-accordion-component="Accordion"
      >
        <div
          aria-expanded="true"
          class="header headerOpen"
          data-accordion-component="AccordionItemButton"
          role="button"
          tabindex="0"
        >
          title
          <div
            class="headerArrow"
          >
            <svg
              aria-hidden="true"
              class="svgIcon inherit"
              data-test-id="ChevronDownIcon"
              focusable="false"
              style="font-size: 24px;"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="collapse body"
          data-accordion-component="AccordionItemPanel"
          role="region"
        >
           You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="accordion accordionOpen"
      data-accordion-component="Accordion"
    >
      <div
        aria-expanded="true"
        class="header headerOpen"
        data-accordion-component="AccordionItemButton"
        role="button"
        tabindex="0"
      >
        title
        <div
          class="headerArrow"
        >
          <svg
            aria-hidden="true"
            class="svgIcon inherit"
            data-test-id="ChevronDownIcon"
            focusable="false"
            style="font-size: 24px;"
            viewBox="0 0 24 24"
          >
            <path
              clip-rule="evenodd"
              d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <div
        class="collapse body"
        data-accordion-component="AccordionItemPanel"
        role="region"
      >
         You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Accordion Snapshots tests should render accordion title with HTML tag 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="accordion"
        data-accordion-component="Accordion"
      >
        <div
          aria-expanded="false"
          class="header"
          data-accordion-component="AccordionItemButton"
          role="button"
          tabindex="0"
        >
          <h1>
            title
          </h1>
          <div
            class="headerArrow"
          >
            <svg
              aria-hidden="true"
              class="svgIcon inherit"
              data-test-id="ChevronDownIcon"
              focusable="false"
              style="font-size: 24px;"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="collapse body"
          data-accordion-component="AccordionItemPanel"
          role="region"
          style="height: 0px;"
        >
           You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="accordion"
      data-accordion-component="Accordion"
    >
      <div
        aria-expanded="false"
        class="header"
        data-accordion-component="AccordionItemButton"
        role="button"
        tabindex="0"
      >
        <h1>
          title
        </h1>
        <div
          class="headerArrow"
        >
          <svg
            aria-hidden="true"
            class="svgIcon inherit"
            data-test-id="ChevronDownIcon"
            focusable="false"
            style="font-size: 24px;"
            viewBox="0 0 24 24"
          >
            <path
              clip-rule="evenodd"
              d="M12.0711 14.7279L5.70718 8.36392L4.29297 9.77814L12.0711 17.5563L19.8493 9.77814L18.4351 8.36392L12.0711 14.7279Z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <div
        class="collapse body"
        data-accordion-component="AccordionItemPanel"
        role="region"
        style="height: 0px;"
      >
         You’ll struggle to find any objective reasons to buy a Q3 Sportback over the standard Audi Q3. But, if you want to stand out, the Sportback is
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
